### PR TITLE
fix: tighten route intent selection

### PR DESCRIPTION
**Pull request by venikman**
## What

Fixes hosted/runtime route over-selection where generic acceptance-check, owner, CI/deploy-health, or pattern-review wording could push compact queries into route:boundary-burden.

## Why

Daily product feedback in discussions #61, #69, and #74 showed project kickoff/project review/spec-writer prompts selecting the boundary route even when exact docs and route reads were healthy. This separates generic acceptance checks from boundary obligations and keeps route:boundary-burden for concrete API, contract, protocol, CI gate, deploy promise, or acceptance-clause review.

## Changes

- Tightened boundary route intent signals and negative API-contract disambiguation.
- Added spec-writer intent for route:writing-or-reviewing-patterns.
- Let explicit route-intent candidates win when route intent is clear.
- Clarified PR/code-review docs so ordinary code review does not force the FPF pattern-writing route.
- Refreshed published/current for the compiler fingerprint.

## Validation

- bun run spec:download
- bun run publish:current
- bun run validate:published
- bun run check
- bun run lint
- bun run check:boundaries
- bun run docs:build
- bun run test tests/query-contracts.test.ts tests/fpf-spec-runtime.test.ts tests/mcp-server.test.ts
- git diff --check
- CLI E2E recording: /Users/stas-studio/.codex/worktrees/5d37/fpf-memory/.runtime/e2e-recordings/2026-05-05T13-12-38Z-route-intent-cli/terminal-recording.txt
- CLI E2E report: /Users/stas-studio/.codex/worktrees/5d37/fpf-memory/.runtime/e2e-recordings/2026-05-05T13-12-38Z-route-intent-cli/report.md

## Caveats

The repo still emits GitHub's existing >50 MB warning for published/current/fpf-index/snapshot.json. The old e2e:report harness is not present on current main, so the E2E evidence for this PR is a terminal recording plus report under .runtime/e2e-recordings.

---

<a href="https://app.devin.ai/review/venikman/fpf-memory/pull/85" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

---

> [!NOTE]
> **Medium Risk**
> Changes the query seeding/ranking heuristics that decide which FPF route wins, which can shift runtime retrieval results for many prompts. Risk is mitigated by added/updated contract tests, but mis-scoring could still cause route regressions in edge cases.
> 
> **Overview**
> Reduces false positives for `route:boundary-burden` by tightening boundary-review seed terms (e.g., focusing on *acceptance clauses / deploy promises* vs generic “acceptance check”) and adding explicit negation handling so prompts can opt out of API/contract review routing.
> 
> Improves route selection when users explicitly ask for a route by letting strong *adoption-route-intent* signals win, and adds new adoption intent scoring for `route:writing-or-reviewing-patterns` to support spec-writer/pattern-edit prompts.
> 
> Updates reviewer-facing docs to recommend starting from the PR packet and only using routes when warranted, refreshes published snapshot metadata (`published/current/*`), and adjusts/adds tests to lock in the new routing behavior (project kickoff/review vs boundary review vs spec writing).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 979bf767f980f317384af629e94048d82b7f84dd. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/runtime/query-helpers.ts [hunk 1]
```diff
@@ -195,36 +208,42 @@ function scoreProjectAlignmentIntent(normalizedQuestion: string): number {
...
   return 0;
 }
 
+function scoreWritingOrReviewingPatternIntent(normalizedQuestion: string): number {
+  if (
+    WRITING_OR_REVIEWING_PATTERN_SIGNALS.some((signal) =>
+      normalizedQuestion.includes(signal),
+    )
+  ) {
+    return 88;
```
[the remark this answers is not in the document]
Comment (venikman) [reply src/runtime/query-helpers.ts:236]: Addressed in 5aea8d5b0ccf386218ab0439edb4d2c9ed37accf by bypassing the compact route fast path whenever the prompt contains exact FPF IDs, plus a regression for `As a spec writer, what is A.1.1?`. Focused local verification passed in a temp PR tree: `bun run test tests/query-contracts.test.ts tests/fpf-spec-runtime.test.ts tests/mcp-server.test.ts` (67/67), `bun run check`, `bun run lint`, `bun run check:boundaries`, `bun run validate:published`. PR CI run 25498284049 is green.

File: src/runtime/route-intent-signals.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  'checking',
+  'check',
+] as const;
+
+export const WRITING_OR_REVIEWING_PATTERN_SIGNALS = [
+  'spec writer',
+  'spec writing',
+  'fpf pattern',
+  'new fpf pattern',
+  'writing reviewing route',
```
[the remark this answers is not in the document]
Comment (venikman) [reply src/runtime/route-intent-signals.ts]: Addressed in 5aea8d5b0ccf386218ab0439edb4d2c9ed37accf by replacing the malformed signal with `writing/reviewing route` and adding `writing or reviewing route`.

File: docs/mcp-recipes.md [hunk 3]
```diff
@@ -93,8 +93,9 @@ Use only the fpf_memory MCP server plus the PR diff, local files, and CI evidenc
 
 Good retrieval shape:
 
-- Start with `query_fpf_spec` for `route:writing-or-reviewing-patterns`.
-- Add `route:boundary-burden` only when the change touches an API, contract, workflow, protocol, CI gate, or deploy promise.
+- For ordinary code review, start from the local diff, CI evidence, and the PR/code review packet rather than forcing a route.
+- Use `route:writing-or-reviewing-patterns` only when the PR changes FPF pattern or specification text.
+- Add `route:boundary-burden` only when the change touches an API, contract, workflow contract, protocol, CI gate, or deploy promise.
```
[the remark this answers is not in the document]
Comment (venikman) [reply docs/mcp-recipes.md]: Addressed in 5aea8d5b0ccf386218ab0439edb4d2c9ed37accf: the docs sentence now says `API, contract, workflow, protocol, CI gate, or deploy promise`.

**Comment by venikman** [comment]
Daily feedback review follow-up, 2026-05-07T13:22Z

Classification: fixed in PR, pending merge and hosted deploy.

Action taken:
- Added commit `5aea8d5b` to address the latest actionable review comments.
- Exact FPF IDs now bypass compact route fast-path selection, so mixed prompts like `As a spec writer, what is A.1.1?` preserve exact-ID grounding.
- Fixed the malformed `writing/reviewing route` signal and corrected the MCP Recipes boundary-route wording.

Validation:
- Temp PR tree: `bun run spec:download`, `bun run publish:current`, `bun run validate:published`, `bun run test tests/query-contracts.test.ts tests/fpf-spec-runtime.test.ts tests/mcp-server.test.ts` (67/67), `bun run check`, `bun run lint`, `bun run check:boundaries`.
- GitHub CI run 25498284049 passed, including validate published, lint/boundaries, typecheck, all four test shards, docs/MCP/Vercel-origin dry-run, Vercel preview, and aggregate `ci`.

Remaining uncertainty:
- Hosted production MCP remains on `main` until this PR merges and production deploy completes.
- Product-role feedback packet routing in #86 is still a separate design choice, not fixed here.

File: src/runtime/route-intent-signals.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  'deploy gate',
+  'deployment gate',
+  'deploy promise',
+  'deployment promise',
+  'slo',
+  'sla',
+  'acceptance clause',
+  'compliance text',
+  'compliance requirement',
+  'compliance',
```
Comment (venikman) [inline src/runtime/route-intent-signals.ts]: **Bare `compliance` keeps over-selecting `route:boundary-burden`** (revives the still-open Bugbot finding from commit `979bf76`).

The PR's stated tightening replaced bare `acceptance` → `acceptance clause` and dropped bare `deploy`, but `compliance` is still listed alongside the narrower `compliance text` / `compliance requirement`. Because `BOUNDARY_BURDEN_SIGNALS` is consumed via `String.includes()` in both the seed rule (`compiler.ts:248`) and `scoreBoundaryBurdenIntent` (`query-helpers.ts:215-220`), the bare term swallows the narrower variants.

Verified locally against `5aea8d5` with a direct `engine.trace(q, 'compact')` call (snapshot loaded from `published/current/fpf-index/snapshot.json`):

| Question | Result |
| --- | --- |
| `Please review this compliance documentation for our team.` | `route:boundary-burden`, score 92, fast-route wins |
| `specification compliance review` | `route:boundary-burden`, score 92 |
| `What is the compliance check process?` | `route:boundary-burden`, score 92 |

Each of those should land on a generic compliance pattern or just stay grounding-level — none of them touches an API/contract/CI gate.

Suggested fix: drop bare `compliance` and rely on the two narrower variants. If a single-word fallback is desired, gate it (e.g., require both `compliance` and a `contract`/`policy`/`api` co-signal).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/runtime/route-intent-signals.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  'reviewer',
+  'reviewing',
+  'checking',
+  'check',
+] as const;
+
+export const WRITING_OR_REVIEWING_PATTERN_SIGNALS = [
+  'spec writer',
+  'spec writing',
+  'fpf pattern',
```
Comment (venikman) [inline src/runtime/route-intent-signals.ts]: **Bare `fpf pattern` still misroutes generic pattern lookups.**

The earlier codex P1 (https://github.com/venikman/fpf-memory/pull/85#discussion_r3188698701) about `fpf pattern` as an 88-score writing-route trigger was partially mitigated in `5aea8d5` by the `extractIds` fast-route bypass at `query-engine.ts:245`, but that bypass only fires when the prompt contains an explicit FPF ID. Plain natural-language lookups still hit the regression:

```
Q: Tell me about FPF patterns.
  routeWins: true | selected: ["route:writing-or-reviewing-patterns", "E.8", "E.19"]
  fastRoute: route:writing-or-reviewing-patterns | score: 88 | reason: adoption-route-intent
```

The user is asking about FPF patterns generally, not asking to author/review one. Combined with the still-open `authoring conventions` / `pattern quality gate(s)` thread on line 64 (https://github.com/venikman/fpf-memory/pull/85#discussion_r3201739060), I'd suggest dropping all four pattern-content phrases (`fpf pattern`, `new fpf pattern`, `authoring conventions`, `pattern quality gate(s)`) from `WRITING_OR_REVIEWING_PATTERN_SIGNALS` and keeping only the explicit authoring-route phrases (`spec writer`, `spec writing`, `writing/reviewing route`, `writing or reviewing route`, `writing or reviewing patterns`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/runtime/query-helpers.ts [hunk 6]
```diff
@@ -195,36 +208,42 @@ function scoreProjectAlignmentIntent(normalizedQuestion: string): number {
...
+    (tokens.has('write') ||
+      tokens.has('writing') ||
+      tokens.has('revise') ||
+      tokens.has('revising') ||
+      tokens.has('spec') ||
+      tokens.has('authoring'))
+  ) {
+    return 72;
+  }
+  return 0;
```
Comment (venikman) [inline src/runtime/query-helpers.ts]: **Second-tier matcher is too permissive.**

`normalizedQuestion.includes('pattern')` is a substring match, so it fires on any prompt containing `pattern`/`patterns`/`anti-pattern`/etc. Combined with `tokens.has('authoring')`, this misroutes general questions about authoring something that happens to use the word "pattern":

```
Q: What is the pattern for authoring docs?
  routeWins: true | selected: ["route:writing-or-reviewing-patterns", "E.8", "E.19"]
  fastRoute: route:writing-or-reviewing-patterns | score: 72 | reason: adoption-route-intent
```

Two tightening options:

1. Require the `pattern` token explicitly (`tokens.has('pattern') || tokens.has('patterns')`) so substrings like `anti-pattern` don't count.
2. Require an authoring-noun co-occurrence (e.g., `authoring patterns`, `writing a pattern`, `revising patterns`) rather than any combination of `pattern` + a verb anywhere in the prompt.

The same caveat applies to `tokens.has('spec')` here — that one was tightened in `b2ef758` to use token matching (good), but the surrounding `pattern` substring check still leaves the function broader than the prior commit's analysis assumed.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by venikman — COMMENTED** [review]
## Review summary

The intent of this PR is solid and the iteration has clearly responded to prior feedback — centralizing the signal lists into `route-intent-signals.ts`, adding the negation handling, and bypassing the compact fast-route when explicit FPF IDs are present are all good changes. CI is green and the new tests pin the intended behavior.

I verified three concerns locally against the PR head (`5aea8d5`) by loading `published/current/fpf-index/snapshot.json` and calling `engine.trace(q, 'compact')`:

**Still-open / actionable**

1. **Pattern-content phrases inside `WRITING_OR_REVIEWING_PATTERN_SIGNALS` misroute natural-language lookups.** `authoring conventions` (E.8 title) and `pattern quality gate(s)` (E.19 title) are already flagged in the open codex thread (https://github.com/venikman/fpf-memory/pull/85#discussion_r3201739060); `fpf pattern` has the same problem ("Tell me about FPF patterns." → `route:writing-or-reviewing-patterns` at 88). The `extractIds` bypass only mitigates queries that include explicit IDs — see inline comment on line 57.

2. **Bare `compliance` undermines the boundary-route tightening** that this PR is built around. Bugbot raised this on `979bf76` and it was never resolved; verified still reproduces ("Please review this compliance documentation for our team." → `route:boundary-burden` at 92). See inline comment on line 28.

3. **Second-tier writing-pattern fallback in `scoreWritingOrReviewingPatternIntent`** is broader than its tests imply, because it uses `normalizedQuestion.includes('pattern')` as a substring rather than a token check. "What is the pattern for authoring docs?" → `route:writing-or-reviewing-patterns` at 72. See inline comment on `query-helpers.ts:232-244`.

**Already addressed in this iteration (good)**

- Duplicate `hasBoundaryReviewNegation` removed via centralization.
- Module-level constants instead of per-call array allocations.
- Malformed `'writing reviewing route'` corrected to `'writing/reviewing route'` + slash-free variant.
- Docs `workflow` typo / duplicate `contract` cleaned up.
- Compact fast-route bypassed for prompts containing explicit FPF IDs.
- `spec` substring guarded behind `tokens.has('spec')` token check.

None of the remaining issues are merge-blocking architectural problems — they're tightenings consistent with this PR's own thesis. Happy to look at a follow-up commit.

---
_Generated by [Claude Code](https://claude.ai/code)_
